File: 0946-validate-stack-sequences/0946-validate-stack-sequences.cpp
class Solution {
public:
    bool validateStackSequences(vector<int>& pushed, vector<int>& popped) {
        stack<int> s;
        int i = 0, j = 0;
        int n = pushed.size();

        while(i < n) {
            s.push(pushed[i]);
            while(s.size() > 0 && s.top() == popped[j]) {
                s.pop();
                j++;
            }
            if(s.size() > 0 && i == n-1) {
                return false;
            }
            i++;
        }

        return true;
    }
};

File: 0946-validate-stack-sequences/0946-validate-stack-sequences.java
class Solution {
    public boolean validateStackSequences(int[] pushed, int[] popped) {
        int i = 0, j = 0;
        int n = pushed.length;

        for(int k  = 0; k < n; k++) {
            pushed[i] = pushed[k];
            i++;
            while(i > 0 && pushed[i-1] == popped[j]) {
                i--;
                j++;
            }
            if(i > 0 && k == n - 1) {
                return false;
            }
        }

        return true;
    }
}